### DISCUSSION TITLE
DS3231 not updating from system time

### DISCUSSION THREAD

**Discussion by grampastever** [micropython/micropython#10856] [RP2040 / Pico] [answered]
I am using Peter Hinch's DS3231 library to set time and get time. While the system time is available from my laptop, the DS3231 time is not being updated. See the attached image. I have tried with multiple DS3231 modules with the same result. Any thoughts on how I can resolve this? 
![DS3231](https://user-images.githubusercontent.com/97183531/221369946-63697628-862f-43cf-98cc-cdc854848c95.png)

**Comment by davefes** [comment, marked as answer]
Have these boards got a memory chip on them as well?

**Reply by grampastever** [reply]
Hi Dave,
thanks for the quick reply. Yes, the board (ZS-042) has an AT24C32N EEPROM on it. I did a fair bit more troubleshooting since I sought help and found that the "troublemaker" appears to be a GY-521 I2C accelerometer shield I have in my circuit (id 104 on the I2C scan). When I remove it from the circuit, the DS3231 works as expected. When I reinstall it, it doesn't. The GY-521 had been working normally in the circuit before I installed the DS3231 so I know that's hooked up correctly. I also have a 64x128 I2C OLED in the circuit (id 60 on the I2C scan) and it works correctly and does not impede operation of the DS3231 and also works with the GY-521. So, it looks like there is a weird kind of interaction going on. 

As you can see from the screenshot in my original post, I had all three devices on I2C bus 1. To solve my problem, I decided to move the DS3231 to the I2C bus 0. And it worked. The DS3231 updates time as expected. But I found a new and interesting potential problem. When I scan the I2C bus 0, both the DS3231 and the GY-521 are seen [87, 104], even though the GY-521 is not on that bus, and when I scan I2C bus 1, the OLED and the GY-521 are seen [60, 104] as expected. At this point, without any further input from someone a lot smarter on the issues I am seeing, I suspect my pico board might have some issues. I'll try a new board to see if the problem persists. In the meantime, I am going to call this problem solved, from a DS3231/library perspective. Thanks for coming to help.

Cheers,

Steve

**Reply by davefes** [reply]
I was tricked by the schematic that was suppose to be for my boards with the memory chip.  Although mine seemed to work there were no pull-up resistors (~10K) on SDA and SCL.  

Good luck!

**Reply by grampastever** [reply]
Hi Dave, I just figured out the whole issue. The DS3231 has the same I2C address as the GY-521 and obviously their data conflicts. The solution is a dollop of solder across a pair of address pads on the back of the DS3231 shield...or use the second I2C bus as I have just done. Overall, there's nothing wrong with any of the components or the software. Something to watch for in the future.

Cheers,

Steve

**Reply by peterhinch** [reply]
I don't think the DS3231 chip supports alternative addresses.

**Reply by davefes** [reply]
I suspect the address pins are for the memory chip, they are on the boards I have.

**Reply by grampastever** [reply]
Hi Peter/Dave, really good pickup. Although I have yet to find a proper system description for the ZS-042 DS3231 RTC module, I have the AT24C32 EEPROM and it makes sense that the module's address pads relate to that chip, not the DS3231 as I had originally thought. Thanks.
